### PR TITLE
feat: Add raw priority queue implementation and update Makefile.

### PR DESCRIPTION
**Pull request by puchkovki**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,4 @@
 .vscode/
 .idea/
-
+.gitignore
```
Comment (trmigor) [inline .gitignore]: Why?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
+REPO_PATH = ~/Github/distr-model
+
+.PHONY: build
+	build: bin/distr-model
```
Comment (trmigor) [inline Makefile]: Superfluous indentation.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+.PHONY: test
+test:
+	@echo "===> $@"
+	@go test ${GOTESTFLAGS} `go list ./...`
+
+.PHONY: clean
+clean:
+	@echo "===> $@"
+	@rm -rf bin
```
Comment (trmigor) [inline Makefile]: No newline at the end of the file.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,22 @@
+REPO_PATH = ~/Github/distr-model
```
Comment (trmigor) [inline Makefile:1]: Makefile should be in a separate commit.

File: priorityq/priorityq.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
+package priorityq
+
+import (
+	"container/heap"
+	"sync"
+
+	model "github.com/puchkovki/distr-model/model"
```
Comment (trmigor) [inline priorityq/priorityq.go]: The package should be independent. It should be just a priority queue, without any relations with the `package model`. Otherwise, there will be a cyclical import.

File: priorityq/priorityq.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+// update modifies the priority and value of an Item in the queue.
+func (pq *PriorityQueue) update(item *Item, value string, priority int) {
+	item.value = value
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}
+
+// A MessageQueue is a PriorityQueue structure with mutex for parallel usage
+type MessageQueue struct {
```
Comment (trmigor) [inline priorityq/priorityq.go]: `MessageQueue` should be in the `package model`, in separate file and in separate commit, not here.

File: priorityq/priorityq.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// Push adds items to the PriorityQueue
+func (pq *PriorityQueue) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Item)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+// Pop removes the highest-priority item from the PriorityQueue
+func (pq *PriorityQueue) Pop() interface{} {
```
Comment (trmigor) [inline priorityq/priorityq.go:45]: As I see, that's just a stack now. There isn't any item redistribution in the slice, so we just get **FILO**. And you don't use the `PriorityQueue` implementing the `heap.Interface`. Check `func main` (lines 77 and 84) in _godoc_ example for `container/heap`: https://play.golang.org/p/fa-K96WIgVS.

File: priorityq/priorityq.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+func (pq *PriorityQueue) update(item *Item, value string, priority int) {
+	item.value = value
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}
+
+// A MessageQueue is a PriorityQueue structure with mutex for parallel usage
+type MessageQueue struct {
+	queue PriorityQueue
+	mutex sync.Mutex
```
Comment (trmigor) [inline priorityq/priorityq.go]: Is there any difference between `std::mutex` and `std::recursive_mutex`? If it is, is it eligible to use simple mutex in this case? If it isn't acceptable, is there any _Golang_ analogue of `std::recursive_mutex`?

File: priorityq/priorityq.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	mq.queue.Pop()
+
+	return ret
+}
+
+func (mq *MessageQueue) enqueue(msg *model.Message) {
+	mq.mutex.Lock()
+	defer mq.mutex.Unlock()
+
+	mq.queue.Push(msg)
```
Comment (trmigor) [inline priorityq/priorityq.go]: This isn't a working use of the priority queue. Check my comment for line 48.

File: priorityq/priorityq_test.go [hunk 10]
```diff
@@ -0,0 +1 @@
+package priorityq
```
Comment (trmigor) [inline priorityq/priorityq_test.go]: Tests have to check whether the priority queue works in the right way. So, every method should be checked un separate test function. Note that you shouldn't check them as `pq.Pop()` or something like that, use `heap.Pop(&pq)`.

**Review by trmigor — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (puchkovki) [reply .gitignore]: Okey (Stirlitz. I'd like you to stay...)

File: Makefile (hunk 2 above)
Comment (puchkovki) [reply Makefile]: Fixed

File: Makefile (hunk 3 above)
Comment (puchkovki) [reply Makefile]: Fixed

File: Makefile (hunk 4 above)
Comment (puchkovki) [reply Makefile:1]: Understand

File: priorityq/priorityq.go (hunk 8 above)
Comment (puchkovki) [reply priorityq/priorityq.go]: Leave std::mutex

File: priorityq/priorityq.go (hunk 5 above)
Comment (puchkovki) [reply priorityq/priorityq.go]: Fixed
